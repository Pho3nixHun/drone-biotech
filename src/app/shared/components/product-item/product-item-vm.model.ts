import { WithId } from '@interfaces/with-id.interace';
import { WithImage } from '@interfaces/with-image.interface';

export interface ProductItemVM extends WithImage, WithId {
  titleKey: string;
  descriptionKey: string;
}
