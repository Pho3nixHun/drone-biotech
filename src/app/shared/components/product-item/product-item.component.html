@if (vm(); as vm) {
  <a
    *transloco="let t"
    class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
  >
    <img [src]="vm.imageSrc" [alt]="vm.titleKey" class="w-auto h-auto object-cover rounded-md" />
    <h3 class="font-bold text-xl sm:text-2xl mt-4">{{ t(vm.titleKey) }}</h3>
    <p class="mt-4">{{ t(vm.descriptionKey) }}</p>
  </a>
}
