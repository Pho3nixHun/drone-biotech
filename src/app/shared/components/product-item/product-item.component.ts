import { Component, Input } from '@angular/core';
import { RouterLink } from '@angular/router';

@Component({
  selector: 'app-product-item',
  standalone: true,
  imports: [RouterLink],
  templateUrl: './product-item.component.html',
})
export class ProductItemComponent {
  @Input() imageSrc!: string;
}
