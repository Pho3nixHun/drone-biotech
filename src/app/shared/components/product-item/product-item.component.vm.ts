export interface ProductItemComponentVM {
  title: string;
  description: string;
  imageSrc: string;
}
