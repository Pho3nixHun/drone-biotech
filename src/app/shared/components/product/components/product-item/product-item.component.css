/* stylelint-disable selector-pseudo-element-no-unknown */
:host ::ng-deep [slot='id'] {
  @apply mb-2;
}

:host ::ng-deep img {
  @apply w-auto h-auto object-cover rounded-md;
}

:host ::ng-deep h3 {
  @apply font-bold text-xl sm:text-2xl mt-4;
}

:host ::ng-deep [slot='description'] {
  @apply mt-4;
}

app-product-item {
  @apply h-full;
}
