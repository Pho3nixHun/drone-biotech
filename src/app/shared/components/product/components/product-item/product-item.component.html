<a
  class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
>
  <ng-content select="[slot=id]" class="" />
  <ng-content select="img" />
  <ng-content select="h3" />
  <ng-content select="[slot=description]" />
</a>
