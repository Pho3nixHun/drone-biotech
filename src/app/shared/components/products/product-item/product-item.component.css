app-product-item {
  @apply h-full;
}
