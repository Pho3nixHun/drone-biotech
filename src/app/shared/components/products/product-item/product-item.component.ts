import { Component, input } from '@angular/core';
import { ProductItemComponentVM } from './product-item.component.vm';

/**
 * ProductItemComponent
 *
 * Type: Presentational
 *
 * Scope:
 * - Responsible for rendering an individual Product Item.
 * - Displays the product's image, name, price, and other relevant details.
 * - Manages the styling and layout necessary for the product's presentation.
 *
 * Out-of-Scope:
 * - Should not handle data fetching.
 * - Contains no business logic related to the product.
 * - Does not manage or transform product data.
 *
 * Purpose (optional):
 * To visually present an individual product.
 */

@Component({
  selector: 'app-product-item',
  standalone: true,
  imports: [],
  templateUrl: './product-item.component.html',
  styleUrl: './product-item.component.css',
})
export class ProductItemComponent {
  vm = input<ProductItemComponentVM | null>(null, { alias: 'vm' });
}
