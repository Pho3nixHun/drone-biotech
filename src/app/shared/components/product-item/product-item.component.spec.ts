import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ProductItemComponent } from './product-item.component';
import { ProductItemVM } from './product-item-vm.model';
import { getTranslocoModule } from 'transloco-testing.module';

const en = {
  title: 'Controller',
  description:
    'The sleek sports car roared to life, its engine purring with power as it sped down the highway.',
};

describe('ProductItemComponent', () => {
  let component: ProductItemComponent;
  let fixture: ComponentFixture<ProductItemComponent>;
  let compiled: HTMLElement;
  const vm: ProductItemVM = {
    id: 1,
    titleKey: 'Controller',
    descriptionKey:
      'The sleek sports car roared to life, its engine purring with power as it sped down the highway.',
    imageSrc: 'assets/lepke.jpg',
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        ProductItemComponent,
        getTranslocoModule({
          langs: { en },
          translocoConfig: { availableLangs: ['en'], defaultLang: 'en' },
        }),
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(ProductItemComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.debugElement.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should get the inputs in the <p>, <img>, <h3> elements', () => {
    fixture.componentRef.setInput('vm', vm);
    fixture.detectChanges();

    const imgElement: HTMLElement | null = compiled.querySelector('img');
    expect(imgElement?.getAttribute('src')).toBe(vm.imageSrc);

    const h3Element: HTMLElement | null = compiled.querySelector('h3');
    expect(h3Element?.innerText).toBe(en.title);

    const pElement: HTMLElement | null = compiled.querySelector('p');
    expect(pElement?.innerText).toBe(en.description);
  });
});
