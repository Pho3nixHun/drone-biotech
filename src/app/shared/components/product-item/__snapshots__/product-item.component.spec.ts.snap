// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItemComponent should render the template correctly when there is a VM provided 1`] = `
<div
  class="h-full"
  id="root0"
  ng-version="18.2.7"
>
  <a
    class="cursor-pointer block p-4 w-full h-full rounded-md shadow-md hover:shadow-xl transition-all"
  >
    <img
      alt="Lepke"
      class="w-auto h-auto object-cover rounded-md"
      src="assets/lepke.jpg"
    />
    <h3
      class="font-bold text-xl sm:text-2xl mt-4"
    >
      Controller
    </h3>
    <p
      class="mt-4"
    >
      The sleek sports car roared to life, its engine purring with power as it sped down the highway.
    </p>
  </a>
  <!--container-->
</div>
`;
